### PR TITLE
persist: better compaction policy

### PR DESCRIPTION
**Pull request by ruchirK**
The first commit is Dan's changes to metrics to make this easier to debug and understand.
The second commit changes the compaction level assignment to only increment the compaction level when the new batch is larger than both of its parents. This change much more tightly bounds the size of the trace to the unsealed size.

Screen Shot 2021-08-24 at 1 19 45 AM

^this graph is from just the first commit. Write amplification improves with this too, specifically the spikes in trace writes are less bad. This change doesn't affect the steady state write amplification

The third commit changes runtime to only drain unsealed to trace once per second. This should let us retain our fast response latencies while easing up on the write pressure downstream. This significantly improved the write amplification at steady state and at peak. 

I don't have a graph or good numbers immediately on hand but running on 6 workers on my laptop the baseline input is ~130kbs and I was able to observe a steady state write rate of ~1.2 mbs for everything. In contrast, the steady state write rate before this change was more like 3.5 - 4 mbs on my laptop.

### CODE REVIEW

File: src/persist/src/indexed/trace.rs [hunk 1]
```diff
@@ -289,6 +289,16 @@ impl Trace {
         // TODO: actually clear the unwanted batches from the blob storage
         let size_bytes = blob.set_trace_batch(key.clone(), new_batch)?;
 
+        // Only upgrade the compaction level if we know this new batch represents
+        // an increase in data over both of its parents so that we we know we need
```
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:293]: nit: "we we"

File: src/persist/src/indexed/trace.rs [hunk 2]
```diff
@@ -289,6 +289,16 @@ impl Trace {
...
+        // even more additional batches to amortize the cost of compacting it in
+        // the future.
+        let merged_level = if size_bytes > first.size_bytes && size_bytes > second.size_bytes {
+            first.level + 1
+        } else {
+            first.level
```
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:299]: I think it's worth adding a test for this. Compaction logic tests will quickly get unwieldy, but this is still straightforward enough.

Eventually, def not now, we should brainstorm how to write a more generalized compaction test framework that tests it at a higher level (e.g. model the mz_metrics workload and assert on how many batches come in before stuff cancels out/assert an upper bound on the total size of the trace). Then we can model different workloads with that and test new compactions strategies vs all of them

**Review by danhhz — APPROVED** [review]
the `persist: only increment compaction levels if the compacted is larger …
…than parents` commit lgtm. let's split the step interval commit into a new PR, tho, so we can evaluate the impact independently

**Comment by ruchirK** [comment]
As far as I can tell, remaining improvements to the trace compaction policy fall under the following three categories:

- reduce write amplification by making merges wider (instead of binary merges, maybe 10-ary merges). The cost here is that meta will get larger, but the benefit is that three levels is enough to fully collapse down the 300 second compaction interval, and our intermediate state is bound to 1/3 the working set (100 batches)

- appropriately assign incoming unsealed batches a compaction level based on their size. Currently, if there's a hiccup and we receive a 20 mb unsealed batch, we'll assign it level 0, and quickly compact it a few times with adjacent neighbors. Instead, we should assign it to a more reasonable initial level, and roll up all batches at levels less than that one into one batch (basically, the big insert pays for some advance compaction).

- reduce duplicate work in a cascade of merges:

in a situation where we have e.g. compaction levels:
`[11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 0]` -- the final result will be a single batch at level 12, but if we do this one step at a time we will repeatedly write and then delete several times more data than is in the working set. Instead, we could detect cases like this and merge into one batch (from a skim I think spine_fueled does something similar here)

There's probably other stuff I'm missing but I haven't observed any other simpler approaches for the write spikes I've seen.

File: src/persist/src/indexed/trace.rs (hunk 2 above)
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:299]: done + cleaned up some wonky antichain handling leftover from before!

**Comment by ruchirK** [comment]
tftr! merging on green!

**Pull request merged**